File: src/shipyard/targets/warm_pool.py
"""Warm-pool runner reuse across PRs.

A "warm pool" entry lets a subsequent ship within a configurable TTL
skip the expensive pre-stage (clone / sync / deps) by re-entering the
same runner workdir that was left in a known-good state by a recent
PASS. Validate (configure / build / test) is always re-run, so the
feature trades a small, bounded correctness risk (stale workdir on the
same SHA) for a potentially large time saving on SSH and cloud runners.

Design points — mirrors issue #82:

- Opt-in per target via ``warm_keepalive_seconds = <N>`` on the target
  section (default 0 = feature off). See
  :func:`extract_warm_keepalive_seconds`.
- Global kill switch: ``SHIPYARD_NO_WARM_POOL=1`` in the env. See
  :func:`warm_reuse_disabled_by_env`.
- Per-ship CLI flag: ``--no-warm`` passed in by the dispatcher.
- GitHub-hosted cloud targets are silently ineligible (workflow runs
  are ephemeral) and surface a single warn-once message so users can
  reconcile the config. See :func:`is_backend_eligible`.
- Same-SHA reuse only — an entry is consulted only if the current job
  SHA matches the stored SHA. The feature is not a cross-SHA cache; it
  only exists to amortise pre-stage cost for back-to-back ships of the
  exact same commit.
- Any failure during a warm reuse evicts the entry so the pool never
  serves a dirty workdir twice.

The pool state is a single JSON file under the machine-global state
directory. Atomic writes (``<path>.tmp`` → rename) keep concurrent
Shipyard processes from corrupting it.
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any

DEFAULT_POOL_FILENAME = "warm_pool.json"

# Backends whose runner identity survives a shipyard invocation.
# Everything else (cloud/github-hosted) is ephemeral per workflow run.
_ELIGIBLE_BACKENDS = {"ssh", "ssh-windows", "local"}


@dataclass(frozen=True)
class PoolEntry:
    """A single warm-pool record.

    Attributes:
        target: Target name the entry belongs to.
        host: Host identifier (SSH alias / ``user@host`` / ``local``).
            Used as the tuple key along with ``target`` so a target
            that was last warm on host A won't be mis-served when a
            subsequent config switches it to host B.
        backend: Backend the entry was warmed on (``ssh`` /
            ``ssh-windows`` / ``local``).
        workdir: Remote repository path the pre-stage was performed in.
            The executor re-enters this directory instead of re-cloning.
        sha: The exact SHA the pre-stage was run against. Reuse is
            only honored when the next job is on this same SHA.
        expires_at: Monotonic-clock-safe UNIX epoch seconds. Entries
            past this point are ignored and pruned.
        created_at: UNIX epoch seconds when the entry was recorded,
            for observability on ``shipyard targets warm status``.
    """

    target: str
    host: str
    backend: str
    workdir: str
    sha: str
    expires_at: float
    created_at: float

    def is_expired(self, *, now: float | None = None) -> bool:
        """Return True if ``now`` (default: wall clock) is past expiry."""
        return (now if now is not None else time.time()) >= self.expires_at

    def ttl_remaining_secs(self, *, now: float | None = None) -> float:
        """Seconds until expiry; clamped to 0 when past-expiry."""
        gap = self.expires_at - (now if now is not None else time.time())
        return max(0.0, gap)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> PoolEntry:
        return cls(
            target=str(data["target"]),
            host=str(data["host"]),
            backend=str(data["backend"]),
            workdir=str(data["workdir"]),
            sha=str(data["sha"]),
            expires_at=float(data["expires_at"]),
            created_at=float(data["created_at"]),
        )


class WarmPool:
    """JSON-backed persistent warm-pool store.

    Not a process-wide singleton — callers construct one against a
    specific state directory. The store is keyed by ``(target, host)``
    so different physical hosts running the same logical target don't
    collide; the first writer wins per key.
    """

    def __init__(self, path: Path) -> None:
        self.path = path

    # ---- load / save -----------------------------------------------

    def _load_raw(self) -> list[dict[str, Any]]:
        """Read the raw list of entries from disk.

        Returns an empty list when the file is missing, unreadable, or
        malformed. The pool is best-effort: a corrupt file must not
        block a ship, so we swallow errors here and let the next save
        overwrite them.
        """
        if not self.path.exists():
            return []
        try:
            payload = json.loads(self.path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return []
        if not isinstance(payload, dict):
            return []
        entries = payload.get("entries", [])
        if not isinstance(entries, list):
            return []
        return [item for item in entries if isinstance(item, dict)]

    def all_entries(self) -> list[PoolEntry]:
        """Return every entry currently in the pool (expired included)."""
        out: list[PoolEntry] = []
        for raw in self._load_raw():
            try:
                out.append(PoolEntry.from_dict(raw))
            except (KeyError, TypeError, ValueError):
                continue
        return out

    def save_entries(self, entries: list[PoolEntry]) -> None:
        """Atomically rewrite the pool file with the given entries."""
        self.path.parent.mkdir(parents=True, exist_ok=True)
        payload = {"entries": [entry.to_dict() for entry in entries]}
        tmp = self.path.with_suffix(self.path.suffix + ".tmp")
        tmp.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
        os.replace(tmp, self.path)

    # ---- lookups ---------------------------------------------------

    def get(self, target: str, host: str) -> PoolEntry | None:
        """Return the unexpired entry for ``(target, host)`` or None.

        Expired entries are ignored here but not pruned; call
        :meth:`prune_expired` if you want the file to shrink.
        """
        now = time.time()
        for entry in self.all_entries():
            if entry.target == target and entry.host == host:
                if entry.is_expired(now=now):
                    return None
                return entry
        return None

    # ---- mutations -------------------------------------------------

    def upsert(self, entry: PoolEntry) -> None:
        """Insert or replace the record for ``(target, host)``."""
        existing = [
            e for e in self.all_entries()
            if not (e.target == entry.target and e.host == entry.host)
        ]
        existing.append(entry)
        self.save_entries(existing)

    def evict(self, target: str, host: str) -> bool:
        """Remove the entry for ``(target, host)``. Return True if removed."""
        kept: list[PoolEntry] = []
        removed = False
        for entry in self.all_entries():
            if entry.target == target and entry.host == host:
                removed = True
                continue
            kept.append(entry)
        if removed:
            self.save_entries(kept)
        return removed

    def drain(self) -> int:
        """Remove every entry. Return count drained."""
        current = self.all_entries()
        self.save_entries([])
        return len(current)

    def prune_expired(self) -> int:
        """Remove expired entries; return how many were pruned."""
        now = time.time()
        kept: list[PoolEntry] = []
        pruned = 0
        for entry in self.all_entries():
            if entry.is_expired(now=now):
                pruned += 1
                continue
            kept.append(entry)
        if pruned:
            self.save_entries(kept)
        return pruned


# ── helpers used by executor / CLI ───────────────────────────────


def default_pool_path(state_dir: Path) -> Path:
    """Canonical location for the warm-pool JSON file."""
    return Path(state_dir) / DEFAULT_POOL_FILENAME


def warm_reuse_disabled_by_env(env: dict[str, str] | None = None) -> bool:
    """Return True when ``SHIPYARD_NO_WARM_POOL=1`` is set.

    Accepts any of ``1|true|yes|on`` (case-insensitive) to match how
    other Shipyard env flags are parsed.
    """
    source = env if env is not None else os.environ
    raw = source.get("SHIPYARD_NO_WARM_POOL", "")
    return raw.strip().lower() in {"1", "true", "yes", "on"}


def extract_warm_keepalive_seconds(target_config: dict[str, Any]) -> int:
    """Return the configured keepalive in seconds; 0 means "off".

    Tolerant of strings / negatives — anything non-positive maps to 0
    so a misconfigured target silently keeps the safe default.
    """
    raw = target_config.get("warm_keepalive_seconds", 0)
    try:
        secs = int(raw)
    except (TypeError, ValueError):
        return 0
    return secs if secs > 0 else 0


def is_backend_eligible(backend: str, target_config: dict[str, Any]) -> bool:
    """Whether ``backend`` is a type where reuse makes sense.

    Cloud + GitHub-hosted runners are excluded because each workflow
    run gets a fresh VM; there's nothing to keep warm. SSH / local
    runners keep the workdir across invocations, which is where the
    whole point of the pool lives.

    ``target_config`` is accepted for future per-target overrides
    (e.g. a Namespace self-managed pool that does persist) but is
    otherwise unused.
    """
    del target_config  # reserved for future overrides
    return backend.lower() in _ELIGIBLE_BACKENDS


def warm_host_key(target_config: dict[str, Any]) -> str:
    """Pick a stable host key for pool indexing.

    SSH backends use their ``host`` field; local backends collapse to
    a literal ``"local"`` so the same machine doesn't spawn dozens of
    identical entries if hostname lookups vary.
    """
    host = target_config.get("host")
    if isinstance(host, str) and host.strip():
        return host.strip()
    return "local"


def compute_expires_at(keepalive_seconds: int, *, now: float | None = None) -> float:
    """Return the absolute expiry time for a new pool entry."""
    base = now if now is not None else time.time()
    return base + max(0, int(keepalive_seconds))


__all__ = [
    "DEFAULT_POOL_FILENAME",
    "PoolEntry",
    "WarmPool",
    "compute_expires_at",
    "default_pool_path",
    "extract_warm_keepalive_seconds",
    "is_backend_eligible",
    "warm_host_key",
    "warm_reuse_disabled_by_env",
]

File: src/warm_pool.rs
//! Warm-pool runner reuse state.
//!
//! The warm pool is an opt-in cache of runner workdirs keyed by
//! `(target, host)`. It only reuses entries for the exact same SHA and
//! only for backends whose workdir survives a Shipyard invocation.

use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};

/// Canonical warm-pool JSON filename.
pub const DEFAULT_POOL_FILENAME: &str = "warm_pool.json";

const ELIGIBLE_BACKENDS: [&str; 3] = ["ssh", "ssh-windows", "local"];

/// A single warm-pool record.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct PoolEntry {
    /// Target name.
    pub target: String,
    /// Host key, normally SSH host or `local`.
    pub host: String,
    /// Backend that warmed the workdir.
    pub backend: String,
    /// Runner workdir path.
    pub workdir: String,
    /// Exact SHA this entry is valid for.
    pub sha: String,
    /// UNIX epoch seconds when the entry expires.
    pub expires_at: f64,
    /// UNIX epoch seconds when the entry was created.
    pub created_at: f64,
}

impl PoolEntry {
    /// Construct a warm-pool entry.
    #[must_use]
    pub fn new(
        target: impl Into<String>,
        host: impl Into<String>,
        backend: impl Into<String>,
        workdir: impl Into<String>,
        sha: impl Into<String>,
        expires_at: f64,
        created_at: f64,
    ) -> Self {
        Self {
            target: target.into(),
            host: host.into(),
            backend: backend.into(),
            workdir: workdir.into(),
            sha: sha.into(),
            expires_at,
            created_at,
        }
    }

    /// Return true if `now` is at or past expiry.
    #[must_use]
    pub fn is_expired(&self, now: f64) -> bool {
        now >= self.expires_at
    }

    /// Return seconds until expiry, clamped at zero.
    #[must_use]
    pub fn ttl_remaining_secs(&self, now: f64) -> f64 {
        (self.expires_at - now).max(0.0)
    }
}

/// JSON-backed persistent warm-pool store.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct WarmPool {
    path: PathBuf,
}

impl WarmPool {
    /// Construct a store at an explicit path.
    #[must_use]
    pub fn new(path: impl Into<PathBuf>) -> Self {
        Self { path: path.into() }
    }

    /// Return the backing JSON path.
    #[must_use]
    pub fn path(&self) -> &Path {
        &self.path
    }

    /// Return every valid entry currently in the pool, including expired entries.
    #[must_use]
    pub fn all_entries(&self) -> Vec<PoolEntry> {
        self.load_raw()
            .into_iter()
            .filter_map(|value| serde_json::from_value(value).ok())
            .collect()
    }

    /// Atomically rewrite the pool file with the given entries.
    pub fn save_entries(&self, entries: &[PoolEntry]) -> Result<(), std::io::Error> {
        if let Some(parent) = self.path.parent() {
            fs::create_dir_all(parent)?;
        }
        let payload = serde_json::json!({ "entries": entries });
        let tmp = self.path.with_extension(format!(
            "{}tmp",
            self.path
                .extension()
                .and_then(|extension| extension.to_str())
                .map_or_else(String::new, |extension| format!("{extension}."))
        ));
        fs::write(
            &tmp,
            serde_json::to_string_pretty(&payload).expect("pool serializes") + "\n",
        )?;
        fs::rename(tmp, &self.path)
    }

    /// Return the unexpired entry for `(target, host)`, if any.
    #[must_use]
    pub fn get(&self, target: &str, host: &str, now: f64) -> Option<PoolEntry> {
        self.all_entries()
            .into_iter()
            .find(|entry| entry.target == target && entry.host == host)
            .filter(|entry| !entry.is_expired(now))
    }

    /// Insert or replace the record for `(target, host)`.
    pub fn upsert(&self, entry: PoolEntry) -> Result<(), std::io::Error> {
        let mut entries = self
            .all_entries()
            .into_iter()
            .filter(|existing| existing.target != entry.target || existing.host != entry.host)
            .collect::<Vec<_>>();
        entries.push(entry);
        self.save_entries(&entries)
    }

    /// Remove the entry for `(target, host)`. Returns true if removed.
    pub fn evict(&self, target: &str, host: &str) -> Result<bool, std::io::Error> {
        let entries = self.all_entries();
        let original_len = entries.len();
        let kept = entries
            .into_iter()
            .filter(|entry| entry.target != target || entry.host != host)
            .collect::<Vec<_>>();
        let removed = kept.len() != original_len;
        if removed {
            self.save_entries(&kept)?;
        }
        Ok(removed)
    }

    /// Remove every entry. Returns count drained.
    pub fn drain(&self) -> Result<usize, std::io::Error> {
        let count = self.all_entries().len();
        self.save_entries(&[])?;
        Ok(count)
    }

    /// Remove expired entries. Returns count pruned.
    pub fn prune_expired(&self, now: f64) -> Result<usize, std::io::Error> {
        let entries = self.all_entries();
        let mut pruned = 0;
        let kept = entries
            .into_iter()
            .filter(|entry| {
                if entry.is_expired(now) {
                    pruned += 1;
                    false
                } else {
                    true
                }
            })
            .collect::<Vec<_>>();
        if pruned > 0 {
            self.save_entries(&kept)?;
        }
        Ok(pruned)
    }

    fn load_raw(&self) -> Vec<serde_json::Value> {
        let Ok(text) = fs::read_to_string(&self.path) else {
            return Vec::new();
        };
        let Ok(payload) = serde_json::from_str::<serde_json::Value>(&text) else {
            return Vec::new();
        };
        payload
            .get("entries")
            .and_then(serde_json::Value::as_array)
            .cloned()
            .unwrap_or_default()
            .into_iter()
            .filter(serde_json::Value::is_object)
            .collect()
    }
}

/// Canonical location for the warm-pool JSON file.
#[must_use]
pub fn default_pool_path(state_dir: &Path) -> PathBuf {
    state_dir.join(DEFAULT_POOL_FILENAME)
}

/// Return true when a warm-pool env value disables reuse.
#[must_use]
pub fn warm_reuse_disabled_by_env_value(value: Option<&str>) -> bool {
    value.unwrap_or_default().trim().eq_ignore_ascii_case("1")
        || matches!(
            value
                .unwrap_or_default()
                .trim()
                .to_ascii_lowercase()
                .as_str(),
            "true" | "yes" | "on"
        )
}

/// Extract `warm_keepalive_seconds`; invalid or non-positive values map to zero.
#[must_use]
pub fn extract_warm_keepalive_seconds(value: Option<&toml::Value>) -> u32 {
    let Some(value) = value else {
        return 0;
    };
    let parsed = value.as_integer().or_else(|| {
        value
            .as_str()
            .and_then(|raw| raw.trim().parse::<i64>().ok())
    });
    parsed
        .and_then(|seconds| u32::try_from(seconds).ok())
        .unwrap_or(0)
}

/// Whether `backend` is a type where reuse makes sense.
#[must_use]
pub fn is_backend_eligible(backend: &str) -> bool {
    let normalized = backend.trim().to_ascii_lowercase().replace('_', "-");
    ELIGIBLE_BACKENDS.contains(&normalized.as_str())
}

/// Pick a stable host key for pool indexing.
#[must_use]
pub fn warm_host_key(host: Option<&str>) -> String {
    host.map(str::trim)
        .filter(|host| !host.is_empty())
        .map_or_else(|| "local".to_owned(), ToOwned::to_owned)
}

/// Return the absolute expiry time for a new pool entry.
#[must_use]
pub fn compute_expires_at(keepalive_seconds: u32, now: f64) -> f64 {
    now + f64::from(keepalive_seconds)
}

/// Return current UNIX epoch seconds.
#[must_use]
pub fn now_epoch_secs() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_or(0.0, |duration| duration.as_secs_f64())
}

/// Render entries keyed by `(target, host)` for status surfaces.
#[must_use]
pub fn entries_by_key(entries: &[PoolEntry]) -> BTreeMap<(String, String), PoolEntry> {
    entries
        .iter()
        .cloned()
        .map(|entry| ((entry.target.clone(), entry.host.clone()), entry))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::{
        PoolEntry, WarmPool, compute_expires_at, default_pool_path, entries_by_key,
        extract_warm_keepalive_seconds, is_backend_eligible, warm_host_key,
        warm_reuse_disabled_by_env_value,
    };

    fn entry(target: &str, host: &str, sha: &str, expires_at: f64) -> PoolEntry {
        PoolEntry::new(target, host, "ssh", "/repo", sha, expires_at, 10.0)
    }

    fn assert_close(actual: f64, expected: f64) {
        assert!(
            (actual - expected).abs() < f64::EPSILON,
            "{actual} != {expected}"
        );
    }

    #[test]
    fn pool_entry_expiry_and_ttl_match_python_contract() {
        let record = entry("ubuntu", "vm", "abc", 20.0);
        assert!(!record.is_expired(19.9));
        assert!(record.is_expired(20.0));
        assert_close(record.ttl_remaining_secs(15.0), 5.0);
        assert_close(record.ttl_remaining_secs(25.0), 0.0);
    }

    #[test]
    fn warm_pool_missing_or_corrupt_file_reads_empty() {
        let temp = tempfile::tempdir().expect("tempdir");
        let pool = WarmPool::new(temp.path().join("warm_pool.json"));
        assert!(pool.all_entries().is_empty());

        std::fs::write(pool.path(), "{not-json").expect("write");
        assert!(pool.all_entries().is_empty());
    }

    #[test]
    fn warm_pool_roundtrips_and_filters_malformed_entries() {
        let temp = tempfile::tempdir().expect("tempdir");
        let pool = WarmPool::new(temp.path().join("state").join("warm_pool.json"));
        pool.save_entries(&[entry("ubuntu", "vm", "abc", 100.0)])
            .expect("save");
        assert_eq!(pool.all_entries().len(), 1);
        assert_eq!(pool.get("ubuntu", "vm", 50.0).expect("entry").sha, "abc");

        std::fs::write(
            pool.path(),
            r#"{"entries":[{"target":"bad"},42,{"target":"mac","host":"local","backend":"local","workdir":"/repo","sha":"def","expires_at":100,"created_at":10}]}"#,
        )
        .expect("write malformed");
        let entries = pool.all_entries();
        assert_eq!(entries.len(), 1);
        assert_eq!(entries[0].target, "mac");
    }

    #[test]
    fn upsert_replaces_by_target_and_host() {
        let temp = tempfile::tempdir().expect("tempdir");
        let pool = WarmPool::new(temp.path().join("warm_pool.json"));
        pool.upsert(entry("ubuntu", "vm", "old", 100.0))
            .expect("upsert old");
        pool.upsert(entry("ubuntu", "vm", "new", 100.0))
            .expect("upsert new");
        pool.upsert(entry("ubuntu", "other", "other", 100.0))
            .expect("upsert other");

        let entries = pool.all_entries();
        assert_eq!(entries.len(), 2);
        assert_eq!(pool.get("ubuntu", "vm", 1.0).expect("entry").sha, "new");
    }

    #[test]
    fn get_ignores_expired_without_pruning() {
        let temp = tempfile::tempdir().expect("tempdir");
        let pool = WarmPool::new(temp.path().join("warm_pool.json"));
        pool.save_entries(&[entry("ubuntu", "vm", "abc", 5.0)])
            .expect("save");

        assert!(pool.get("ubuntu", "vm", 5.0).is_none());
        assert_eq!(pool.all_entries().len(), 1);
    }

    #[test]
    fn evict_drain_and_prune_mutate_store() {
        let temp = tempfile::tempdir().expect("tempdir");
        let pool = WarmPool::new(temp.path().join("warm_pool.json"));
        pool.save_entries(&[
            entry("ubuntu", "vm", "abc", 5.0),
            entry("mac", "local", "def", 50.0),
        ])
        .expect("save");

        assert_eq!(pool.prune_expired(10.0).expect("prune"), 1);
        assert_eq!(pool.all_entries()[0].target, "mac");
        assert!(pool.evict("mac", "local").expect("evict"));
        assert!(!pool.evict("mac", "local").expect("evict missing"));
        pool.upsert(entry("ubuntu", "vm", "abc", 100.0))
            .expect("upsert");
        assert_eq!(pool.drain().expect("drain"), 1);
        assert!(pool.all_entries().is_empty());
    }

    #[test]
    fn helper_contracts_match_python_defaults() {
        let temp = tempfile::tempdir().expect("tempdir");
        assert_eq!(
            default_pool_path(temp.path()),
            temp.path().join("warm_pool.json")
        );
        for value in ["1", "true", "TRUE", "yes", "on"] {
            assert!(warm_reuse_disabled_by_env_value(Some(value)));
        }
        assert!(!warm_reuse_disabled_by_env_value(Some("0")));

        assert_eq!(
            extract_warm_keepalive_seconds(Some(&toml::Value::Integer(30))),
            30
        );
        assert_eq!(
            extract_warm_keepalive_seconds(Some(&toml::Value::String("45".to_owned()))),
            45
        );
        assert_eq!(
            extract_warm_keepalive_seconds(Some(&toml::Value::Integer(-1))),
            0
        );
        assert_eq!(
            extract_warm_keepalive_seconds(Some(&toml::Value::String("bad".to_owned()))),
            0
        );

        assert!(is_backend_eligible("ssh"));
        assert!(is_backend_eligible("ssh_windows"));
        assert!(is_backend_eligible("local"));
        assert!(!is_backend_eligible("cloud"));
        assert_eq!(warm_host_key(Some(" vm ")), "vm");
        assert_eq!(warm_host_key(None), "local");
        assert_close(compute_expires_at(30, 10.0), 40.0);
    }

    #[test]
    fn entries_by_key_keeps_latest_for_key() {
        let map = entries_by_key(&[
            entry("ubuntu", "vm", "old", 10.0),
            entry("ubuntu", "vm", "new", 20.0),
        ]);
        assert_eq!(map.len(), 1);
        assert_eq!(map[&("ubuntu".to_owned(), "vm".to_owned())].sha, "new");
    }
}
